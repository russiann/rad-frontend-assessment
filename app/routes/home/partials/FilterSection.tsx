import React from "react";
import { Button, Dropdown, DropdownTrigger, DropdownMenu, DropdownItem } from "@heroui/react";
import { Icon } from "@iconify/react";

type FilterSectionProps = {
  selectedCategory: string;
  selectedSort: string;
  onCategoryFilter: (category: string) => void;
  onSortChange: (sort: string) => void;
};

export function FilterSection({ 
  selectedCategory, 
  selectedSort, 
  onCategoryFilter, 
  onSortChange 
}: FilterSectionProps) {
  return (
    <div className="flex flex-col sm:flex-row gap-4 mb-6">
      <Dropdown>
        <DropdownTrigger>
          <Button 
            variant="bordered" 
            endContent={<Icon icon="lucide:chevron-down" className="text-sm" />}
          >
            Filter by Category
          </Button>
        </DropdownTrigger>
        <DropdownMenu 
          aria-label="Categories"
          onAction={(key) => onCategoryFilter(key as string)}
        >
          <DropdownItem key="all">All Categories</DropdownItem>
          <DropdownItem key="electronics">Electronics</DropdownItem>
          <DropdownItem key="clothing">Clothing</DropdownItem>
          <DropdownItem key="footwear">Footwear</DropdownItem>
          <DropdownItem key="home">Home</DropdownItem>
        </DropdownMenu>
      </Dropdown>
      
      <Dropdown>
        <DropdownTrigger>
          <Button 
            variant="bordered" 
            endContent={<Icon icon="lucide:chevron-down" className="text-sm" />}
          >
            Sort by Price
          </Button>
        </DropdownTrigger>
        <DropdownMenu 
          aria-label="Sort options"
          onAction={(key) => onSortChange(key as string)}
        >
          <DropdownItem key="default">Default</DropdownItem>
          <DropdownItem key="low-to-high">Low to High</DropdownItem>
          <DropdownItem key="high-to-low">High to Low</DropdownItem>
        </DropdownMenu>
      </Dropdown>
    </div>
  );
} 