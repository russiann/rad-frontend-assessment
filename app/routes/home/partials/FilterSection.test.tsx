import { describe, it, expect, vi } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FilterSection } from './FilterSection';

describe('FilterSection', () => {
  const mockOnCategoryFilter = vi.fn();
  const mockOnSortChange = vi.fn();

  const defaultProps = {
    selectedCategory: 'all',
    selectedSort: 'default',
    onCategoryFilter: mockOnCategoryFilter,
    onSortChange: mockOnSortChange,
  };

  beforeEach(() => {
    mockOnCategoryFilter.mockClear();
    mockOnSortChange.mockClear();
  });

  describe('Basic Rendering', () => {
    it('should render category and sort buttons', () => {
      render(<FilterSection {...defaultProps} />);

      expect(screen.getByText('Filter by Category')).toBeInTheDocument();
      expect(screen.getByText('Sort by Price')).toBeInTheDocument();
    });
  });

  describe('Category Filtering', () => {
    it('should show category options when clicked', async () => {
      const user = userEvent.setup();
      render(<FilterSection {...defaultProps} />);

      const categoryButton = screen.getByText('Filter by Category');
      await user.click(categoryButton);

      await waitFor(() => {
        expect(screen.getByText('All Categories')).toBeInTheDocument();
        expect(screen.getByText('Electronics')).toBeInTheDocument();
        expect(screen.getByText('Clothing')).toBeInTheDocument();
        expect(screen.getByText('Footwear')).toBeInTheDocument();
        expect(screen.getByText('Home')).toBeInTheDocument();
      });
    });

    it('should call onCategoryFilter when category is selected', async () => {
      const user = userEvent.setup();
      render(<FilterSection {...defaultProps} />);

      const categoryButton = screen.getByText('Filter by Category');
      await user.click(categoryButton);

      const electronicsOption = screen.getByText('Electronics');
      await user.click(electronicsOption);

      expect(mockOnCategoryFilter).toHaveBeenCalledWith('electronics');
    });

    it('should call onCategoryFilter with "all" for All Categories', async () => {
      const user = userEvent.setup();
      render(<FilterSection {...defaultProps} />);

      const categoryButton = screen.getByText('Filter by Category');
      await user.click(categoryButton);

      const allCategoriesOption = screen.getByText('All Categories');
      await user.click(allCategoriesOption);

      expect(mockOnCategoryFilter).toHaveBeenCalledWith('all');
    });
  });

  describe('Price Sorting', () => {
    it('should show sort options when clicked', async () => {
      const user = userEvent.setup();
      render(<FilterSection {...defaultProps} />);

      const sortButton = screen.getByText('Sort by Price');
      await user.click(sortButton);

      await waitFor(() => {
        expect(screen.getByText('Default')).toBeInTheDocument();
        expect(screen.getByText('Low to High')).toBeInTheDocument();
        expect(screen.getByText('High to Low')).toBeInTheDocument();
      });
    });

    it('should call onSortChange when sort option is selected', async () => {
      const user = userEvent.setup();
      render(<FilterSection {...defaultProps} />);

      const sortButton = screen.getByText('Sort by Price');
      await user.click(sortButton);

      const lowToHighOption = screen.getByText('Low to High');
      await user.click(lowToHighOption);

      expect(mockOnSortChange).toHaveBeenCalledWith('low-to-high');
    });

    it('should call onSortChange with correct values for all options', async () => {
      const user = userEvent.setup();
      render(<FilterSection {...defaultProps} />);

      const sortButton = screen.getByText('Sort by Price');

      // Test Default
      await user.click(sortButton);
      await user.click(screen.getByText('Default'));
      expect(mockOnSortChange).toHaveBeenCalledWith('default');

      // Test High to Low
      await user.click(sortButton);
      await user.click(screen.getByText('High to Low'));
      expect(mockOnSortChange).toHaveBeenCalledWith('high-to-low');
    });
  });

  describe('Props Handling', () => {
    it('should handle different selectedCategory values', () => {
      const propsWithElectronics = {
        ...defaultProps,
        selectedCategory: 'electronics',
      };

      render(<FilterSection {...propsWithElectronics} />);
      expect(screen.getByText('Filter by Category')).toBeInTheDocument();
    });

    it('should handle different selectedSort values', () => {
      const propsWithSort = {
        ...defaultProps,
        selectedSort: 'low-to-high',
      };

      render(<FilterSection {...propsWithSort} />);
      expect(screen.getByText('Sort by Price')).toBeInTheDocument();
    });
  });

  describe('snapshots', () => {
    it('renders filter section correctly', () => {
      const { container } = render(<FilterSection {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('renders with electronics category selected', () => {
      const props = { ...defaultProps, selectedCategory: 'electronics' };
      const { container } = render(<FilterSection {...props} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('renders with low-to-high sort selected', () => {
      const props = { ...defaultProps, selectedSort: 'low-to-high' };
      const { container } = render(<FilterSection {...props} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});