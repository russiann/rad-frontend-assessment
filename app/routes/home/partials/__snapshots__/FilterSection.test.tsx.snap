// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilterSection > snapshots > renders filter section correctly 1`] = `
<div
  class="flex flex-col sm:flex-row gap-4 mb-6"
>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r5s»"
    tabindex="0"
    type="button"
  >
    Filter by Category
    <svg
      aria-hidden="true"
      class="iconify iconify--lucide text-sm"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="m6 9l6 6l6-6"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r5v»"
    tabindex="0"
    type="button"
  >
    Sort by Price
    <svg
      aria-hidden="true"
      class="iconify iconify--lucide text-sm"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="m6 9l6 6l6-6"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
</div>
`;

exports[`FilterSection > snapshots > renders with electronics category selected 1`] = `
<div
  class="flex flex-col sm:flex-row gap-4 mb-6"
>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r62»"
    tabindex="0"
    type="button"
  >
    Filter by Category
    <svg
      aria-hidden="true"
      class="iconify iconify--lucide text-sm"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="m6 9l6 6l6-6"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r65»"
    tabindex="0"
    type="button"
  >
    Sort by Price
    <svg
      aria-hidden="true"
      class="iconify iconify--lucide text-sm"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="m6 9l6 6l6-6"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
</div>
`;

exports[`FilterSection > snapshots > renders with low-to-high sort selected 1`] = `
<div
  class="flex flex-col sm:flex-row gap-4 mb-6"
>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r68»"
    tabindex="0"
    type="button"
  >
    Filter by Category
    <svg
      aria-hidden="true"
      class="iconify iconify--lucide text-sm"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="m6 9l6 6l6-6"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r6b»"
    tabindex="0"
    type="button"
  >
    Sort by Price
    <svg
      aria-hidden="true"
      class="iconify iconify--lucide text-sm"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="m6 9l6 6l6-6"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
</div>
`;
